src/runtime/algorithm.go

```diff
@@ -0,0 +1,55 @@
+package runtime
+
+// This file implements various core algorithms used in the runtime package and
+// standard library.
+
+import "unsafe"
+
+// This function is used by hash/maphash.
+func fastrand() uint32 {
+	xorshift32State = xorshift32(xorshift32State)
+	return xorshift32State
+}
+
+var xorshift32State uint32 = 1
+
+func xorshift32(x uint32) uint32 {
+	// Algorithm "xor" from p. 4 of Marsaglia, "Xorshift RNGs".
+	// This could probably be improved if needed.
+	x ^= x << 13
+	x ^= x >> 17
+	x ^= x << 5
+	return x
+}
+
+// This function is used by hash/maphash.
+func memhash(p unsafe.Pointer, seed, s uintptr) uintptr {
+	if unsafe.Sizeof(uintptr(0)) > 4 {
+		return seed ^ uintptr(hash64(p, s))
+	}
+	return seed ^ uintptr(hash32(p, s))
+}
+
+// Get FNV-1a hash of the given memory buffer.
+//
+// https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function#FNV-1a_hash
```

Comment:
I should lookup the state-of-the-art for embedded hash functions.  (I dislike fnv, but that's just my personal bias...). Also, I know from experiments that the issue with the maps isn't the hashing, it's the bucket chains.

Reply:
This is not new code, I'm only moving the code to a shared file.

Reply:
Yeah, this was more areas of research for my own interests if it becomes a bottleneck.